File: scripts/rust/review_import_order.py
#!/usr/bin/env python3
"""
Review: Import order.

RustRules.md Line 50: "Import order: after the module declaration add a blank line,
then all use std::… lines, then a blank line, then use statements from external crates,
then another blank line followed by use crate::Types::Types::*; if needed and the rest
of the internal crate::… imports."

RustRules.md Lines 75-86: "Inside src/ use crate::, outside src/ (tests/benches) use apas_ai::"

Checks:
1. Import ordering: std → external → internal (crate:: or apas_ai::)
2. Blank lines between sections
3. crate:: in src/ vs apas_ai:: in tests/benches/
"""
# Git commit: 584a672b6a34782766863c5f76a461d3297a741a
# Date: 2025-10-17 05:17:36 -0700


import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "lib"))
from review_utils import ReviewContext, create_review_parser


def check_file_imports(file_path, repo_root):
    """Check if imports follow the correct order."""
    with open(file_path, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    # Determine if file is in src/, tests/, or benches/
    relative_path = file_path.relative_to(repo_root)
    in_src = relative_path.parts[0] == 'src'
    in_tests_benches = relative_path.parts[0] in ('tests', 'benches')
    
    # Find first use statement
    first_use_idx = None
    for idx, line in enumerate(lines):
        stripped = line.strip()
        if stripped.startswith('use '):
            first_use_idx = idx
            break
    
    if first_use_idx is None:
        return []  # No imports, no problem
    
    violations = []
    
    # Track import sections
    std_section = []
    external_section = []
    internal_section = []
    
    # Parse imports
    i = first_use_idx
    current_section = None
    blank_after_std = False
    blank_after_external = False
    
    while i < len(lines):
        line = lines[i]
        stripped = line.strip()
        
        # Stop at first non-import, non-blank, non-comment line
        if not stripped.startswith('use ') and stripped and not stripped.startswith('//'):
            break
        
        if stripped.startswith('use std::') or stripped.startswith('use core::') or stripped.startswith('use alloc::'):
            if current_section == 'external':
                violations.append((i + 1, "std import after external imports", line.strip()))
            elif current_section == 'internal':
                violations.append((i + 1, "std import after internal imports", line.strip()))
            std_section.append((i + 1, line))
            current_section = 'std'
            
        elif stripped.startswith('use crate::') or stripped.startswith('use apas_ai::'):
            # Check crate:: vs apas_ai::
            if in_src and stripped.startswith('use apas_ai::'):
                violations.append((i + 1, "use apas_ai:: in src/ (should be crate::)", line.strip()))
            elif in_tests_benches and stripped.startswith('use crate::'):
                violations.append((i + 1, "use crate:: in tests/benches (should be apas_ai::)", line.strip()))
            
            if current_section == 'external' and not blank_after_external:
                violations.append((i + 1, "missing blank line before internal imports", line.strip()))
            
            internal_section.append((i + 1, line))
            current_section = 'internal'
            
        elif stripped.startswith('use ') and not stripped.startswith('use self::') and not stripped.startswith('use super::'):
            # External crate import
            if current_section == 'std' and not blank_after_std:
                violations.append((i + 1, "missing blank line before external imports", line.strip()))
            if current_section == 'internal':
                violations.append((i + 1, "external import after internal imports", line.strip()))
                
            external_section.append((i + 1, line))
            current_section = 'external'
            
        elif not stripped:  # Blank line
            if current_section == 'std' and not blank_after_std:
                blank_after_std = True
            elif current_section == 'external' and not blank_after_external:
                blank_after_external = True
        
        i += 1
    
    # Check for required blank lines
    if std_section and external_section and not blank_after_std:
        violations.append((std_section[-1][0], "missing blank line after std imports", ""))
    if external_section and internal_section and not blank_after_external:
        violations.append((external_section[-1][0], "missing blank line after external imports", ""))
    
    return violations


def main():
    parser = create_review_parser(__doc__)
    args = parser.parse_args()
    context = ReviewContext(args)

    search_dirs = [
        context.repo_root / "src",
        context.repo_root / "tests",
        context.repo_root / "benches",
    ]

    if context.dry_run:
        files = context.find_files(search_dirs)
        print(f"Would check {len(files)} file(s) for import order")
        return 0

    all_violations = []
    files = context.find_files(search_dirs)

    for rust_file in files:
        violations = check_file_imports(rust_file, context.repo_root)
        if violations:
            all_violations.append((rust_file, violations))
    
    if all_violations:
        print("✗ Found import order violations (RustRules.md Lines 50, 75-86):\n")
        for file_path, violations in all_violations:
            rel_path = file_path.relative_to(context.repo_root)
            print(f"  {rel_path}:")
            for line_num, reason, line_content in violations:
                print(f"    Line {line_num}: {reason}")
                if line_content:
                    print(f"      {line_content}")
            print()
        
        total = sum(len(v) for _, v in all_violations)
        print(f"Total violations: {total}")
        print("\nExpected:")
        print("  - Order: std imports → [blank] → external imports → [blank] → internal imports")
        print("  - In src/: use crate::")
        print("  - In tests/benches/: use apas_ai::")
        return 1
    else:
        print("✓ Import order correct: std → external → internal, with blank lines")
        return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/bin/review_import_order.rs
// Copyright (C) Brian G. Milnes 2025

//! Review: Import order
//! 
//! Replaces: scripts/rust/review_import_order.py
//! RustRules.md Line 50: "Import order: after the module declaration add a blank line,
//! then all use std::… lines, then a blank line, then use statements from external crates,
//! then another blank line followed by use crate::Types::Types::*; if needed and the rest
//! of the internal crate::… imports."
//! 
//! RustRules.md Lines 75-86: "Inside src/ use crate::, outside src/ (tests/benches) use apas_ai::"
//! 
//! Binary: rusticate-review-import-order
//!
//! Uses AST parsing to find USE items and check their ordering

use anyhow::Result;
use rusticate::{StandardArgs, format_number, parse_source, find_nodes, find_rust_files};
use ra_ap_syntax::{SyntaxKind, SyntaxNode, ast::{self, AstNode}};
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;


macro_rules! log {
    ($($arg:tt)*) => {{
        use std::io::Write;
        let msg = format!($($arg)*);
        println!("{}", msg);
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open("analyses/review_import_order.log")
        {
            let _ = writeln!(file, "{}", msg);
        }
    }};
}
#[derive(Debug)]
struct Violation {
    file: PathBuf,
    line_num: usize,
    message: String,
    context: String,
}

#[derive(Debug, PartialEq, Eq)]
enum ImportSection {
    Std,      // std::, core::, alloc::
    External, // Other crates
    Internal, // crate:: or apas_ai::
}

fn classify_use_path(use_text: &str) -> ImportSection {
    // Extract the path after "use " and before any trailing tokens
    // e.g. "use crate::Foo;" -> "crate::Foo"
    // e.g. "use std::collections::HashMap;" -> "std::collections::HashMap"
    let trimmed = use_text.trim();
    let path = if let Some(start_idx) = trimmed.find("use ") {
        &trimmed[start_idx + 4..] // Skip "use "
    } else {
        trimmed
    };
    
    // Remove trailing semicolon and whitespace
    let path = path.trim().trim_end_matches(';').trim();
    
    if path.starts_with("std::") || path.starts_with("core::") || path.starts_with("alloc::") {
        ImportSection::Std
    } else if path.starts_with("crate::") || path.starts_with("apas_ai::") {
        ImportSection::Internal
    } else if path.starts_with("self::") || path.starts_with("super::") {
        // self/super are local, treat as internal
        ImportSection::Internal
    } else {
        ImportSection::External
    }
}

fn has_blank_line_between(node1: &SyntaxNode, node2: &SyntaxNode, source: &str) -> bool {
    let end_line = rusticate::line_number(node1, source);
    let start_line = rusticate::line_number(node2, source);
    
    if start_line <= end_line + 1 {
        return false; // Adjacent or same line
    }
    
    // Check if there's an empty line between them
    let lines: Vec<&str> = source.lines().collect();
    for line_idx in end_line..start_line - 1 {
        if let Some(line) = lines.get(line_idx) {
            let trimmed = line.trim();
            // Empty line (ignore comments)
            if trimmed.is_empty() {
                return true;
            }
        }
    }
    false
}

fn check_file(file_path: &Path, source: &str, in_src: bool) -> Result<Vec<Violation>> {
    let source_file = parse_source(source)?;
    let root = source_file.syntax();
    
    // Find all USE items
    let uses = find_nodes(root, SyntaxKind::USE);
    
    if uses.is_empty() {
        return Ok(Vec::new());
    }
    
    let mut violations = Vec::new();
    
    // Track the section we're in and detect transitions
    let mut current_section: Option<ImportSection> = None;
    let mut _seen_types_import = false;
    let mut seen_other_internal = false;
    let mut prev_use: Option<&SyntaxNode> = None;
    
    for use_node in &uses {
        let line_num = rusticate::line_number(use_node, source);
        let use_text = use_node.text().to_string();
        let section = classify_use_path(&use_text);
        
        // Check ordering violations
        match (&current_section, &section) {
            (Some(ImportSection::External), ImportSection::Std) => {
                violations.push(Violation {
                    file: file_path.to_path_buf(),
                    line_num,
                    message: "std import after external imports".to_string(),
                    context: use_text.trim().to_string(),
                });
            }
            (Some(ImportSection::Internal), ImportSection::Std) => {
                violations.push(Violation {
                    file: file_path.to_path_buf(),
                    line_num,
                    message: "std import after internal imports".to_string(),
                    context: use_text.trim().to_string(),
                });
            }
            (Some(ImportSection::Internal), ImportSection::External) => {
                violations.push(Violation {
                    file: file_path.to_path_buf(),
                    line_num,
                    message: "external import after internal imports".to_string(),
                    context: use_text.trim().to_string(),
                });
            }
            _ => {}
        }
        
        // Check for missing blank lines when transitioning sections
        if let (Some(prev_section), Some(prev_node)) = (&current_section, prev_use) {
            if prev_section != &section {
                // Section changed - should have blank line
                if !has_blank_line_between(prev_node, use_node, source) {
                    violations.push(Violation {
                        file: file_path.to_path_buf(),
                        line_num,
                        message: format!("missing blank line between {:?} and {:?} imports", prev_section, section),
                        context: use_text.trim().to_string(),
                    });
                }
            }
        }
        
        // Check crate:: vs apas_ai:: usage using AST
        let (has_apas_ai, has_crate) = if let Some(use_ast) = ast::Use::cast(use_node.clone()) {
            if let Some(use_tree) = use_ast.use_tree() {
                if let Some(path) = use_tree.path() {
                    let first_segment = path.segments().next().map(|s| s.to_string());
                    (first_segment == Some("apas_ai".to_string()), 
                     first_segment == Some("crate".to_string()))
                } else {
                    (false, false)
                }
            } else {
                (false, false)
            }
        } else {
            (false, false)
        };
        
        if has_apas_ai && in_src {
            violations.push(Violation {
                file: file_path.to_path_buf(),
                line_num,
                message: "use apas_ai:: in src/ (should be crate::)".to_string(),
                context: use_text.trim().to_string(),
            });
        } else if has_crate && !in_src {
            violations.push(Violation {
                file: file_path.to_path_buf(),
                line_num,
                message: "use crate:: in tests/benches (should be apas_ai::)".to_string(),
                context: use_text.trim().to_string(),
            });
        }
        
        // Check Types::Types::* ordering within internal section using AST
        if section == ImportSection::Internal {
            let is_types_import = if let Some(use_ast) = ast::Use::cast(use_node.clone()) {
                if let Some(use_tree) = use_ast.use_tree() {
                    if let Some(path) = use_tree.path() {
                        let segments: Vec<String> = path.segments().map(|s| s.to_string()).collect();
                        // Check for Types::Types pattern (consecutive "Types" segments)
                        segments.windows(2).any(|w| w[0] == "Types" && w[1] == "Types")
                    } else {
                        false
                    }
                } else {
                    false
                }
            } else {
                false
            };
            
            if is_types_import {
                if seen_other_internal {
                    // Types import comes after other internal imports
                    violations.push(Violation {
                        file: file_path.to_path_buf(),
                        line_num,
                        message: "use crate::Types::Types::* should come first within internal imports".to_string(),
                        context: use_text.trim().to_string(),
                    });
                }
                _seen_types_import = true;
            } else {
                // This is some other internal import
                seen_other_internal = true;
            }
        }
        
        current_section = Some(section);
        prev_use = Some(use_node);
    }
    
    Ok(violations)
}

fn main() -> Result<()> {
    let start = Instant::now();
    let args = StandardArgs::parse()?;
    let base_dir = args.base_dir();
    
    // Print compilation directory for Emacs compile-mode
    log!("Entering directory '{}'", base_dir.display());
    log!("");
    
    let search_dirs = args.get_search_dirs();
    
    let mut all_violations = Vec::new();
    let mut total_files = 0;
    
    let files = find_rust_files(&search_dirs);
    
    for file in &files {
        total_files += 1;
        let in_src = file.components().any(|c| c.as_os_str() == "src");
        
        match fs::read_to_string(file) {
            Ok(source) => {
                match check_file(file, &source, in_src) {
                    Ok(violations) => all_violations.extend(violations),
                    Err(e) => {
                        eprintln!("Warning: Failed to parse {}: {}", file.display(), e);
                    }
                }
            }
            Err(e) => {
                eprintln!("Warning: Failed to read {}: {}", file.display(), e);
            }
        }
    }
    
    // Report violations
    if all_violations.is_empty() {
        log!("✓ Import order correct: std → external → internal, with blank lines");
    } else {
        log!("✗ Found {} violation(s) (RustRules.md Lines 50, 75-86):", format_number(all_violations.len()));
        log!("");
        for v in &all_violations {
            if let Ok(rel_path) = v.file.strip_prefix(&base_dir) {
                log!("{}:{}: {}", rel_path.display(), v.line_num, v.message);
                log!("  {}", v.context);
            }
        }
    }
    
    // Summary line
    let unique_files: std::collections::HashSet<_> = all_violations.iter().map(|v| &v.file).collect();
    log!("");
    log!("Summary: {} files checked, {} files with violations, {} total violations",
             format_number(total_files), format_number(unique_files.len()), format_number(all_violations.len()));
    
    let elapsed = start.elapsed().as_millis();
    log!("Completed in {}ms", elapsed);
    
    // Exit code: 0 if no violations, 1 if violations found
    if all_violations.is_empty() {
        Ok(())
    } else {
        std::process::exit(1);
    }
}

